raiden/network/matrixtransport.py

```diff
@@ -0,0 +1,722 @@
+        else:
+            # no room with expected name => create one and invite peer
+            address = address_encoder(receiver_address)
+            candidates = self._client.search_user_directory(address)
+            if not candidates and not allow_missing_peers:
+                raise ValueError('No candidates found for given address: {}'.format(address))
+
+            # filter candidates
+            peers = [user for user in candidates if _validate_userid_signature(user)]
+            if not peers and not allow_missing_peers:
+                raise ValueError('No valid peer found for given address: {}'.format(address))
+
+            try:
+                # Try join first to avoid races
+                room_name_full = f'#{room_name}:{self._server_name}'
+                log.trace('Trying to join', room_name=room_name_full)
+                room = self._client.join_room(room_name_full)
+                for user in peers:
+                    room.invite_user(user.user_id)
+            except MatrixRequestError:
+                room = self._client.create_room(
+                    room_name,
+                    invitees=[user.user_id for user in peers],
+                    is_public=True  # FIXME: This is for debugging purposes only
+                )
+            offline_peers = [
+                user for user in peers
+                if self._userid_to_presence.get(user.user_id) is UserPresence.OFFLINE
+            ]
+            if offline_peers:
+                log.warning('Inviting offline peers', offline_peers=offline_peers, room=room)
+
+        room.add_listener(self._handle_message, 'm.room.message')
+        log.info('CHANNEL ROOM', peer_address=address_encoder(receiver_address), room=room)
+        self._address_to_roomid[receiver_address] = room.room_id
+        return room
+
+    def _make_room_alias(self, *parts):
+        return self._room_sep.join([self._room_prefix, self._network_name, *parts])
+
+    def _handle_presence_change(self, event):
+        """
+        Update node network reachability from presence events.
+
+        Due to the possibility of nodes using accounts on multiple homeservers a composite
+        address state is synthesised from the cached individual user presence state.
+        """
+        if event['type'] != 'm.presence':
+            return
+        user_id = event['sender']
+        new_state = UserPresence(event['content']['presence'])
+        if new_state == self._userid_to_presence.get(user_id):
+            return
+
+        log.trace(
+            'Changing user presence state',
+            user_id=user_id,
+            prev_state=self._userid_to_presence.get(user_id),
+            state=new_state
+        )
+        self._userid_to_presence[user_id] = new_state
+
+        # User should be re-validated after presence change
+        self._userids_to_address.pop(user_id, None)
+
+        try:
+            # FIXME: This should probably use ecrecover instead
```

Comment:
@ulope Can you make detailed issues about those fixmes so that you don't forget them?

Reply:
I'm working on fixing it.

Reply:
What I mean is: Instead of parsing the sender address from the matrix username we should use the signature in the displayname to ensure the user is really who he claims to be (although Bernd poked a bit of a hole into that whole mechanism in case of a malicious server operator - I'll cerate an issue for it)

Reply:
Hmm guys can you please synchronize before "fixing" stuff? @andrevmatos As I told you in person any changes you want to do should be done on top of master if this PR is merged. There is no reason to not merge the PR if tests pass.

Reply:
About the malicious server operator, I  don't think it's an issue, as it's just an additional security layer to early-drop messages from other users trying to impersonate someone. If someone connected on a server, it already trusts it at least to receive his signed userId.